<a href="https://colab.research.google.com/github/Sharma-Shivam2021/credit-default-prediction/blob/main/notebooks/03_model_comparison.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
!git clone https://github.com/Sharma-Shivam2021/credit-default-prediction.git
%cd credit-default-prediction

Cloning into 'credit-default-prediction'...
remote: Enumerating objects: 37, done.
remote: Counting objects: 100% (37/37), done.
remote: Compressing objects: 100% (32/32), done.
remote: Total 37 (delta 13), reused 10 (delta 2), pack-reused 0 (from 0)
Receiving objects: 100% (37/37), 602.08 KiB | 4.04 MiB/s, done.
Resolving deltas: 100% (13/13), done.
/content/credit-default-prediction


In [6]:
!git config user.email "2571sharmashivam@gmail.com"
!git config user.name "Shivam Sharma"

In [4]:
import getpass

token = getpass.getpass("Give your token here: ")

username = "Sharma-Shivam2021"
!git remote set-url origin https://{username}:{token}@github.com/{username}/credit-default-prediction.git
!git push origin main

fatal: not a git repository (or any of the parent directories): .git


In [7]:
%cd /content/credit-default-prediction

/content/credit-default-prediction


In [8]:
!python src/download_data.py

Dataset not found. Downloading...
Dataset downloaded and extracted successfully.


# Credit Default Prediction: Model Comparison

## Objective

The objective of this notebook is to compare multiple classification algorithms using a consistent evaluation framework.

To avoid using the test set for model-selection decisions, model comparison will be performed using the training data and cross-validation. The final test set will be reserved for evaluating the selected model.

The comparison will focus on metrics appropriate for the imbalanced credit-default problem, particularly:

- Precision
- Recall
- F1-score
- ROC-AUC
- Average Precision

In [15]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_validate, StratifiedKFold

In [16]:
df = pd.read_excel(
    "data/raw/default of credit card clients.xls",
    header = 1
)

In [17]:
df_processed = df.copy()

In [18]:
df_processed["EDUCATION"] = df_processed["EDUCATION"].replace([0,5,6],4)
df_processed['MARRIAGE'] = df_processed["MARRIAGE"].replace(0,3);

In [19]:
X = df_processed.drop(columns=["ID", "default payment next month"])
y = df_processed["default payment next month"]

In [20]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

In [21]:
cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

In [22]:
scoring = {
    "accuracy": "accuracy",
    "precision": "precision",
    "recall": "recall",
    "f1": "f1",
    "roc_auc": "roc_auc",
    "average_precision": "average_precision"
}

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

In [25]:
categorical_features = [
    "SEX", "EDUCATION", "MARRIAGE",
    "PAY_0", "PAY_2", "PAY_3",
    "PAY_4", "PAY_5", "PAY_6"
]

numerical_features = [
    "LIMIT_BAL", "AGE",
    "BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6",
    "PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6"
]

In [27]:
preprocessor_scaled = ColumnTransformer(
    transformers = [
        ("cat",OneHotEncoder(handle_unknown="ignore"),categorical_features),
        ("num",StandardScaler(),numerical_features)
    ]
)

In [29]:
preprocessor_tree = ColumnTransformer(
    transformers=[
        ("cat",OneHotEncoder(handle_unknown="ignore"),categorical_features),
        ("num","passthrough", numerical_features)
    ]
)

In [30]:
logistic_pipeline = Pipeline(
    steps=[
        ("preprocessor", preprocessor_scaled),
        ("classifier", LogisticRegression(max_iter=1000))
    ]
)

In [31]:
tree_pipeline =  Pipeline(
    steps=[
        ("preprocessor", preprocessor_tree),
        ("classifier", DecisionTreeClassifier(random_state=42))
    ]
)

In [32]:
logistic_cv = cross_validate(
    logistic_pipeline,
    X_train,
    y_train,
    cv=cv,
    scoring=scoring
)

In [38]:
logistic_cv["test_f1"]

array([0.47416413, 0.46687697, 0.47073171, 0.46993865, 0.45962733])

In [39]:
print("Mean F1:", logistic_cv["test_f1"].mean())
print("Std F1:", logistic_cv["test_f1"].std())

Mean F1: 0.4682677584327605
Std F1: 0.004903498158481453


In [40]:
tree_cv = cross_validate(
    tree_pipeline,
    X_train,
    y_train,
    cv=cv,
    scoring=scoring
)

In [41]:
tree_cv["test_f1"]

array([0.39210408, 0.41251133, 0.39890461, 0.40946315, 0.4150772 ])

In [42]:
print("Mean F1:", tree_cv["test_f1"].mean())
print("Std F1:", tree_cv["test_f1"].std())

Mean F1: 0.40561207518083453
Std F1: 0.008711774930910632


In [43]:
tree_cv = cross_validate(
    tree_pipeline,
    X_train,
    y_train,
    cv=cv,
    scoring=scoring,
    return_train_score=True
)

In [44]:
print("Mean Train F1:", tree_cv["train_f1"].mean())
print("Mean Validation F1:", tree_cv["test_f1"].mean())

Mean Train F1: 0.9990101898300887
Mean Validation F1: 0.40561207518083453


### Logistic Regression vs Decision Tree

Using 5-fold stratified cross-validation, Logistic Regression achieved a mean F1-score of approximately **0.4683**, with a standard deviation of **0.0049**.

The unrestricted Decision Tree achieved a lower mean validation F1-score of approximately **0.4056**, with a standard deviation of **0.0087**.

Further investigation showed that the Decision Tree achieved a mean training F1-score of approximately **0.9990**, compared with only **0.4056** on validation folds. This large train-validation performance gap indicates severe overfitting.

Therefore, the unrestricted Decision Tree does not generalize as well as Logistic Regression in its current configuration.

In [45]:
from sklearn.ensemble import RandomForestClassifier

In [46]:
forest_pipeline = Pipeline(
    steps = [
        ("preprocessor", preprocessor_tree),
        ("classifier", RandomForestClassifier(
            random_state=4,
            n_estimators=100,
            n_jobs=-1
            )
        )
    ]
)

In [47]:
forest_cv = cross_validate(
    forest_pipeline,
    X_train,
    y_train,
    cv=cv,
    scoring=scoring,
    return_train_score=True,
    n_jobs=-1
)

In [48]:
print("Mean Train F1:", forest_cv["train_f1"].mean())
print("Mean Validation F1:", forest_cv["test_f1"].mean())
print("Std Validation F1:", forest_cv["test_f1"].std())

Mean Train F1: 0.9989871184708754
Mean Validation F1: 0.47166505169735073
Std Validation F1: 0.008568073105878347


### Random Forest

Random Forest achieved a mean validation F1-score of approximately **0.4717**, improving substantially over the single Decision Tree and slightly exceeding the Logistic Regression baseline of **0.4683**.

However, the mean training F1-score remained approximately **0.9990**, indicating a substantial gap between training and validation performance. Although aggregating multiple trees improved generalization compared with a single unrestricted Decision Tree, the model still fits the training data very closely.

The difference between the validation F1-scores of Random Forest and Logistic Regression is small, so additional evaluation metrics and further model comparison are required before selecting a preferred model.

In [49]:
models_cv = {
    "Logistic Regression": logistic_cv,
    "Decision Tree": tree_cv,
    "Random Forest": forest_cv
}

for name, results in models_cv.items():
    print(f"\n{name}")
    print("Accuracy:", results["test_accuracy"].mean())
    print("Precision:", results["test_precision"].mean())
    print("Recall:", results["test_recall"].mean())
    print("F1:", results["test_f1"].mean())
    print("ROC-AUC:", results["test_roc_auc"].mean())
    print("Average Precision:", results["test_average_precision"].mean())


Logistic Regression
Accuracy: 0.8203333333333334
Precision: 0.6785127047498885
Recall: 0.35769634232708725
F1: 0.4682677584327605
ROC-AUC: 0.7689298041192015
Average Precision: 0.5429635751644949

Decision Tree
Accuracy: 0.7269166666666667
Precision: 0.3911886527676349
Recall: 0.4211698447437038
F1: 0.40561207518083453
ROC-AUC: 0.6175532400005744
Average Precision: 0.2928752328876981

Random Forest
Accuracy: 0.8174166666666667
Precision: 0.6553546249078289
Recall: 0.36843009561743084
F1: 0.47166505169735073
ROC-AUC: 0.7668162904549872
Average Precision: 0.5321193587265924


### Initial Model Comparison

Five-fold stratified cross-validation was used to compare Logistic Regression, Decision Tree, and Random Forest on the same training data.

Logistic Regression achieved the highest mean accuracy, precision, ROC-AUC, and average precision among the three models. Random Forest achieved slightly higher recall and F1-score than Logistic Regression, although the differences were relatively small.

The unrestricted Decision Tree achieved higher recall than the other two models but substantially lower precision, F1-score, ROC-AUC, and average precision. Its near-perfect training F1-score combined with much lower validation performance also indicated severe overfitting.

At this stage, Logistic Regression and Random Forest appear to be the stronger candidates. However, additional models and subsequent hyperparameter tuning should be evaluated before selecting the final model.

In [50]:
from sklearn.neighbors import KNeighborsClassifier

In [52]:
knn_pipeline = Pipeline(
    steps=[
        ("preprocessor", preprocessor_scaled),
        ("classifier", KNeighborsClassifier(n_neighbors=5))
    ]
)

In [ ]:
knn_cv = cross_validate(
    knn_pipeline,
    X_train,
    y_train,
    cv=cv,
    scoring=scoring,
    return_train_score=True,
    n_jobs=-1
)

In [ ]:
print("Mean Train F1:", knn_cv["train_f1"].mean())
print("Mean Validation F1:", knn_cv["test_f1"].mean())
print("Std Validation F1:", knn_cv["test_f1"].std())